napi_basics.md

```diff
@@ -0,0 +1,120 @@
+- All N-API APIs generally return a status code of type `napi_status`. This status 
+can be used to determine whether the API call succeeded or failed.
+- The actual value being returned by the API is returned in an out parameter.
+- All JavaScript objects are abstracted behind an opaque type named `napi_value`.
+- In case of an error status code, additional information can be obtained using 
+`napi_get_last_error_info`
+
+## N-API types
+
+N-API exposes the following fundamental datatypes as abstractions that are consumed by 
+the various APIs. These APIs should be treated as opaque types that can be introspected
+only with other N-API APIs.
+
+### *napi_env*
+
+`napi_env` is used to represent a context that the underlying N-API implementation can use 
+to persist VM specific state.
+<< REVIEW: What else can we say here? >>
+
+### *napi_value*
+This is an opaque pointer that is used to represent a JavaScript object.
+
+### *napi_ref*
+This is the abstraction to use to reference a `napi_value`. This allows for users 
+to manage the lifetimes of JavaScript objects, including defining their minimum
+lifetimes explicitly, and also being notified when they are garbage-collected.
+
+### *napi_handle_scope*
+This is an abstraction used to provide hints to the garbage collector about the location of 
+handles to GC pointers. napi_handle_scope allows the user to provide a variable stack
+separate from the thread stack, in a manner that is efficient for scan in case 
+the underlying VM uses a garbage collector with precise-semantics.
+N-API values are created within the context of a handle scope- if the user does not 
+create a handle scope, they are created within the default handle scope.
+Handle scopes are created using `napi_open_handle_scope` and are destroyed using 
+`napi_close_handle_scope`. Calling close can indicate to the GC that all napi_values 
+created during the lifetime of the handle scope are no longer referenced from the
+current stack frame.
+
+Consider creating a `napi_handle_scope` per function with a non-trivial amount of 
+objects created.
+
+### *napi_escapable_handle_scope*
+Escapable Handle Scopes are a special type of handle scope that are used to return
+values created within a particular handle scope to a parent scope.
+
+### *napi_callback_info*
+Opaque datatype that is passed to a callback function. It can be used for two purposes:
+- Get additional information about the context in which the callback was invoked
+- Set the return value of the callback
+
+### *napi_callback*
+Function pointer type for N-API APIs that call back into user-provided native code.
+Your callback function should satisfy the following signature:
+```
+typedef void (*napi_callback)(napi_env, napi_callback_info);
+```
+
+### *napi_finalize*
+Function pointer type for N-API APIs that allow the user to be notified when an object 
+is garbage collected. The user must provide a function satisfying the following signature
```

Comment:
... allows the user to be notified _when externally-owned data is ready to be cleaned up_ because the object it was associated with is garbage-collected.